BSP/Src/BSP_I2C.c

```diff
@@ -0,0 +1,214 @@
+
+static QueueHandle_t I2C_MetaRegisterQueue;
+
+static QueueHandle_t I2C_ReceiveQueue; 
+
+HAL_StatusTypeDef BSP_I2C_Init() {
+    //QUEUES
+    I2C_DataQueue = xQueueCreateStatic(I2C_QUEUE_SIZE, I2C_ITEM_LENGTH, __DataBufferStorage, &__DataStaticQueue);
+    I2C_DataRegisterQueue = xQueueCreateStatic(I2C_QUEUE_SIZE, I2C_ITEM_LENGTH, __DataBufferStorage, &__DataStaticQueue);
+    I2C_MetaQueue = xQueueCreateStatic(I2C_QUEUE_SIZE * 2, I2C_ITEM_LENGTH, __MetaBufferStorage, &__MetaStaticQueue);
+    I2C_MetaRegisterQueue = xQueueCreateStatic(I2C_QUEUE_SIZE * 2, I2C_ITEM_LENGTH, __MetaBufferStorage, &__MetaStaticQueue);
+    I2C_ReceiveQueue = xQueueCreateStatic(I2C_QUEUE_SIZE, I2C_ITEM_LENGTH, __ReceiveBufferStorage, &__ReceiveStaticQueue);
+
+    semaphore = xSemaphoreCreateBinaryStatic(&semaphoreBuffer);
+
+    GPIO_InitTypeDef gpio_struct;
+    I2C_HandleTypeDef i2c_struct;
+
+    if (__HAL_RCC_I2C3_IS_CLK_DISABLED()) {
+        __HAL_RCC_I2C3_CLK_ENABLE();
+    }
+    if (__HAL_RCC_GPIOA_IS_CLK_DISABLED()) {
+        __HAL_RCC_GPIOA_CLK_ENABLE();
+    }
+    if (__HAL_RCC_GPIOC_IS_CLK_DISABLED()) {
+        __HAL_RCC_GPIOC_CLK_ENABLE();
+    }
+
+    // i2c config 
+    i2c_struct.Init.ClockSpeed = I2C_CLOCK_SPEED;
+    i2c_struct.Init.DutyCycle = I2C_DUTYCYCLE_2;
+    // 7 bit addr mode
+    i2c_struct.Init.OwnAddress1 = 0x50 << 1;
+    i2c_struct.Init.AddressingMode = I2C_ADDRESSINGMODE_7BIT;
+    i2c_struct.Init.DualAddressMode = I2C_DUALADDRESS_DISABLE;
+    i2c_struct.Init.GeneralCallMode = I2C_GENERALCALL_ENABLED; 
+    i2c_struct.Init.NoStretchMode = I2C_NOSTRETCH_DISABLE;
+    HAL_StatusTypeDef stat = HAL_I2C_Init(&i2c_struct);
+
+    i2c_struct.Instance = I2C1;
+    HAL_NVIC_SetPriority(I2C1_EV_IRQn, 0, 0);
+    HAL_NVIC_EnableIRQ(I2C1_EV_IRQn);
+
+    return stat;
+}
+
+/** 
+ * @brief    preform I2C write
+ * @param    hi2c               struct with config data
+ * @param    deviceAddress      target device address
+ * @param    pDataBuff          data buffer
+ * @param    len                amount of data   
+ * @return   void
+ */
+void BSP_I2C_Write(I2C_HandleTypeDef *hi2c, 
+              uint32_t deviceAddress, 
+              uint8_t* pDataBuff, 
+              uint16_t len) {
+    //copy buffer data over for security incase it changes 
+    uint8_t buffer[I2C_QUEUE_SIZE * I2C_ITEM_LENGTH];
+        for (uint16_t i = 0; i < len; i++) {
+            buffer[i] = pDataBuff[i];
+        }
+    xSemaphoreTake(semaphore, 1);
+    if (!__HAL_I2C_GET_FLAG(hi2c, I2C_FLAG_BUSY)) {
+        // HAL requires 7b I2C address left shifted
+        HAL_I2C_Master_Transmit_IT(hi2c, deviceAddress << 1, buffer, len);
+    }
+    else if (uxQueueSpacesAvailable(I2C_DataQueue) >= len) {
+        for (uint32_t i = 0; i < len; i++) {
+            xQueueSend(I2C_DataQueue, buffer[i], 0);
+        }
+            // HAL requires 7b I2C address left shifted
+            xQueueSend(I2C_MetaQueue, deviceAddress << 1, 0);
```

Comment:
alternating pushes to the metaqueue of the address and length, you should make a struct containing address and length and just push those. then the static queue can be initialized as `I2C_QUEUE_SIZE * sizeof(struct)`